werkzeug/wsgi.py

```diff
@@ -764,6 +780,87 @@ def __next__(self):
         raise StopIteration()
 
 
+@implements_iterator
+class RangeWrapper(object):
+
+    """This class can be used to convert an iterable object into
+    an iterable that will only yield a piece of the underlying content.
+    It yields blocks until the underlying stream range is fully read.
+    The yielded blocks will have a size that can't exceed the original
+    iterator defined block size, but that can be smaller.
+
+    If you're using this object together with a :class:`BaseResponse` you have
+    to use the `direct_passthrough` mode.
+
+    :param iterable: an iterable object with a :meth:`__next__` method.
+    :param start_byte: byte from which read will start.
+    :param byte_range: how many bytes to read.
+    """
+
+    def __init__(self, iterable, start_byte=0, byte_range=None):
+        self.iterable = iterable
+        self.byte_range = byte_range
+        self.start_byte = start_byte
+        self.end_byte = None
+        if byte_range is not None:
+            self.end_byte = self.start_byte + self.byte_range
+        self.read_length = 0
+        self.seekable = hasattr(iterable, 'seekable') and iterable.seekable()
+        self.end_reached = False
+
+    def __iter__(self):
+        return self
+
+    def _next_chunk(self):
+        chunk = next(self.iterable)
+        self.read_length += len(chunk)
+        return chunk
+
+    def _first_iteration(self):
+        chunk = None
+        if self.seekable:
+            self.iterable.seek(self.start_byte)
+            self.read_length = self.iterable.tell()
+            contextual_read_length = self.read_length
+        else:
+            for chunk in self.iterable:
```

Comment:
It's late here, but I think a way to shorten this entire for-else construct is:

    while self.read_length <= self.start_byte:
        chunk = self._next_chunk()

    chunk = chunk[self.start_byte - self.read_length:]
    contextual_read_length = self.start_byte

The StopIteration from `_next_chunk` bubbles up naturally. I'm unsure why `end_reached` is needed, but it could be set in `_next_chunk` as well if StopIteration is raised.